node_modules/oae-principals/lib/internal/dao.js

```diff
@@ -211,6 +210,86 @@ var updatePrincipal = module.exports.updatePrincipal = function(principalId, pro
 };
 
 /**
+ * Marks a principal as being deleted
+ *
+ * @param  {String}     principalId     The id of the principal to mark as deleted
+ * @param  {Function}   callback        Standard callback function
+ * @param  {Object}     callback.err    An error that occurred, if any
+ */
+var deletePrincipal = module.exports.deletePrincipal = function(principalId, callback) {
+    var deleted = Date.now();
+
+    // Determine if the principal was already deleted in the authz index before we set them and flip
+    // the principals deleted flag
+    AuthzDelete.isDeleted([principalId], function(err, wasDeleted) {
+        if (err) {
+            return callback(err);
+        }
+
+        // Set (or re-Set) the principal as deleted in the authz index
+        AuthzDelete.setDeleted(principalId, function(err) {
+            if (err) {
+                return callback(err);
+            }
+
+            Cassandra.runQuery('UPDATE "Principals" SET "deleted" = ? WHERE "principalId" = ?', [deleted, principalId], function(err) {
+                if (err) {
+                    // If updating the principal field fails, we want to make sure we try and unset
+                    // the principal as being deleted if it wasn't set before
+                    if (!wasDeleted[principalId]) {
+                        AuthzDelete.unsetDeleted(principalId);
+                    }
+
+                    return callback(err);
+                }
+
+                // Invalidate the principal from the user cache if they are a user
+                return OaeUtil.invokeIfNecessary(isUser(principalId), invalidateCachedUsers, [principalId], callback);
```

Comment:
We don't honour the `deleted` flag for users (`user.model.js` doesn't accept it). I think I should actually remove this and have the DAO give an error if a user is attempted to be deleted until we have actual support for it.